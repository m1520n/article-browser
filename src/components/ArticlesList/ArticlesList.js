import React from 'react';
import PropTypes from 'prop-types';

import { ArticlesListWrapper, NoResults } from './ArticlesList.style';

import ArticleCard from '../ArticleCard/ArticleCard';

const propTypes = {
  articles: PropTypes.arrayOf(
    PropTypes.shape({
      content: PropTypes.string,
      urlToImage: PropTypes.string,
      title: PropTypes.string,
      publishedAt: PropTypes.string,
      author: PropTypes.string,
      source: PropTypes.shape({
        name: PropTypes.string,
      }),
    }),
  ).isRequired,
  isLoading: PropTypes.bool.isRequired,
  handleCTA: PropTypes.func.isRequired,
};

const renderNoResults = isLoading =>
  !isLoading && <NoResults>No Results - try different filter settings</NoResults>;

const ArticlesList = ({ articles, isLoading, handleCTA }) => (
  <ArticlesListWrapper>
    {articles.length
      ? articles.map(article => (
        <ArticleCard
          article={article}
          key={`${article.author}-${article.publishedAt}`}
          handleCTA={handleCTA}
        />
        ))
      : renderNoResults(isLoading)}
  </ArticlesListWrapper>
);

ArticlesList.propTypes = propTypes;

export default ArticlesList;
