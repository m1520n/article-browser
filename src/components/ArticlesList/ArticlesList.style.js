import styled from 'styled-components';

import { fonts, colors } from '../../styles/variables';

export const ArticlesListWrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  margin-bottom: 6.5rem;
`;

export const NoResults = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 80vh;
  font-family: ${fonts.arialMT};
  font-size: 2.4rem;
  color: ${colors.darkSlate};
`;
