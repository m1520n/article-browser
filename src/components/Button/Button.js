import React from 'react';
import PropTypes from 'prop-types';

import { ButtonWrapper } from './Button.style';

const propTypes = {
  type: PropTypes.oneOf(['primary', 'secondary']),
  size: PropTypes.oneOf(['small', 'regular', 'big']),
  disabled: PropTypes.bool,
  centered: PropTypes.bool,
  block: PropTypes.bool,
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,
  handleClick: PropTypes.func,
};

const defaultProps = {
  type: 'primary',
  size: 'regular',
  disabled: false,
  centered: false,
  block: false,
  handleClick: () => {},
};

const Button = ({ children, handleClick, type, size, disabled, block, centered }) => (
  <ButtonWrapper
    type={type}
    size={size}
    disabled={disabled}
    block={block}
    centered={centered}
    onClick={handleClick}
  >
    {children}
  </ButtonWrapper>
);

Button.propTypes = propTypes;
Button.defaultProps = defaultProps;

export default Button;
