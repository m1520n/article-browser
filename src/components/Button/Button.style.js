import styled from 'styled-components';

import { fonts, colors } from '../../styles/variables';

const sizes = {
  small: {
    width: '22.5rem',
    height: '3.2rem',
  },
  regular: {
    width: '33rem',
    height: '4.8rem',
  },
  big: {
    width: '22.5rem',
    height: '6.4rem',
  },
};

export const ButtonWrapper = styled.button`
  border: none;
  cursor: pointer;
  font-family: ${fonts.arialMT};
  font-size: 1.4rem;
  line-height: ${({ size }) => sizes[size].height};
  width: ${({ size }) => sizes[size].width};
  ${({ block }) => block && 'width: 100%;'}
  height: ${({ size }) => sizes[size].height};
  background-color: ${({ type }) => (type === 'primary' ? colors.royalBlue : colors.white)};
  color: ${({ type }) => (type === 'primary' ? colors.white : colors.royalBlue)};
  border: 1px solid ${colors.royalBlue};
  ${({ disabled }) => disabled && 'pointer-events: none;'}

  :hover {
    background-color: ${({ type }) => (type === 'primary' ? colors.sapphire : colors.royalBlue)};
    border: 1px solid ${({ type }) => (type === 'primary' ? colors.sapphire : colors.royalBlue)};
    color: ${colors.white};
  }

  /* TODO: Should it use the outline for accesibility?
  /* :focus {
    outline:0;
  } */
`;

export default {
  ButtonWrapper,
};
