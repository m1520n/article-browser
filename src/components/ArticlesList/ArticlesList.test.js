import React from 'react';
import renderer from 'react-test-renderer';

import ArticlesList from './ArticlesList';
import Button from '../Button/Button';

import { NoResults } from './ArticlesList.style';

const articles = [
  {
    publishedAt: '2018-08-26',
    author: 'Dan Abramov',
    source: {
      name: 'Medium',
    },
  },
  {
    publishedAt: '2019-01-23',
    author: 'Kyle Simpson',
    source: {
      name: 'HackerNoon',
    },
  },
  {
    publishedAt: '2018-03-23',
    author: 'Douglas Crockford',
    source: {
      name: 'Medium',
    },
  },
];

describe('ArticlesList', () => {
  test('should be rendered rendered correctly', () => {
    const component = renderer.create(
      <ArticlesList articles={articles} isLoading={false} handleCTA={() => {}} />,
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should be rendered correctly when there are no articles', () => {
    const component = renderer.create(
      <ArticlesList articles={[]} isLoading={false} handleCTA={() => {}} />,
    );

    expect(component.root.findAllByType(NoResults)).toHaveLength(1);
  });

  test('should invoke handleCTA callback when button is clicked', () => {
    const handleCTAMock = jest.fn();
    const component = renderer.create(
      <ArticlesList articles={articles} isLoading={false} handleCTA={handleCTAMock} />,
    );

    const button = component.root.findAllByType(Button)[0];
    button.props.handleClick();
    expect(handleCTAMock).toHaveBeenCalledWith(articles[0]);
  });
});
