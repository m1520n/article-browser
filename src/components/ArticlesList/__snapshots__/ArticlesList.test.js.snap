// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticlesList should be rendered rendered correctly 1`] = `
<div
  className="sc-bdVaJa glwcuW"
>
  <div
    className="sc-bxivhb dernUS"
  >
    <div
      className="sc-ifAKCX bxjQWs"
    />
    <div
      className="sc-htoDjs cRxafb"
    >
      <span
        className="sc-dnqmqq hBpdmm"
      >
        Aug 26 2018
      </span>
      <span
        className="sc-dnqmqq sc-iwsKbI fFdvtX"
      >
        Dan Abramov
      </span>
      <a
        className="sc-gZMcBi kifsrt"
        rel="noopener noreferrer"
        target="_blank"
      >
        Medium
      </a>
    </div>
    <h2
      className="sc-EHOje bjWfG"
    />
    <p
      className="sc-bZQynM kmgnWI"
    />
    <button
      className="sc-htpNat evLaha"
      disabled={false}
      onClick={[Function]}
      size="regular"
      type="primary"
    >
      Read More
    </button>
  </div>
  <div
    className="sc-bxivhb dernUS"
  >
    <div
      className="sc-ifAKCX bxjQWs"
    />
    <div
      className="sc-htoDjs cRxafb"
    >
      <span
        className="sc-dnqmqq hBpdmm"
      >
        Jan 23 2019
      </span>
      <span
        className="sc-dnqmqq sc-iwsKbI fFdvtX"
      >
        Kyle Simpson
      </span>
      <a
        className="sc-gZMcBi kifsrt"
        rel="noopener noreferrer"
        target="_blank"
      >
        HackerNoon
      </a>
    </div>
    <h2
      className="sc-EHOje bjWfG"
    />
    <p
      className="sc-bZQynM kmgnWI"
    />
    <button
      className="sc-htpNat evLaha"
      disabled={false}
      onClick={[Function]}
      size="regular"
      type="primary"
    >
      Read More
    </button>
  </div>
  <div
    className="sc-bxivhb dernUS"
  >
    <div
      className="sc-ifAKCX bxjQWs"
    />
    <div
      className="sc-htoDjs cRxafb"
    >
      <span
        className="sc-dnqmqq hBpdmm"
      >
        Mar 23 2018
      </span>
      <span
        className="sc-dnqmqq sc-iwsKbI fFdvtX"
      >
        Douglas Crockford
      </span>
      <a
        className="sc-gZMcBi kifsrt"
        rel="noopener noreferrer"
        target="_blank"
      >
        Medium
      </a>
    </div>
    <h2
      className="sc-EHOje bjWfG"
    />
    <p
      className="sc-bZQynM kmgnWI"
    />
    <button
      className="sc-htpNat evLaha"
      disabled={false}
      onClick={[Function]}
      size="regular"
      type="primary"
    >
      Read More
    </button>
  </div>
</div>
`;
